### PR TITLE
Deprecate `*_enabled` flags under `cupy.cuda`

### PR DESCRIPTION
**Pull request by kmaehashi**
Currently when using `*_enabled` flags, users should write:

```py
import cupy
if cupy.cuda.cutensor_enabled:
    from cupy.cuda import cutensor as cuda_cutensor
    from cupy import cutensor

def foo():
    if cupy.cuda.cutensor_enabled:
        # do cutensor things
        ...
```

I think this is better:

```py
# This import is always safe.
# Dummy object is returned when cuTENSOR is unavailable.
from cupy.cuda import cutensor


def foo():
    if cupy.cuda.cutensor.available:
        # do cutensor things
        pass
```

This is related to #3731.
In this way, we can use `__getattr__` in dummy objects to warn users if the library is not installed.
(note: module-level `__getattr__` is another choice, but only available in Python 3.7 or later)

I would like to hear opinions before completing this PR.

### CODE REVIEW

File: cupy/cuda/thrust.pyx [hunk 1]
```diff
@@ -14,6 +14,9 @@ from cupy.cuda cimport stream
 from cupy_backends.cuda.api cimport runtime
 
 
+available = True
```
Comment (leofang) [inline cupy/cuda/thrust.pyx:17]: Could you move this line to around line 69 (in the Python Interface section)?

File: cupy/cuda/__init__.py [hunk 2]
```diff
@@ -14,48 +14,53 @@
...
 from cupy_backends.cuda.libs import profiler  # NOQA
 
 
 _available = None
 
+
+class UnavailableModule():
+    available = False
+
+# All `*_enabled` flags are kept just for backward compatibility.
```
Comment (leofang) [inline cupy/cuda/__init__.py:29]: 👍 for keeping the old flags around. I wonder if it's possible to emit a warning when users attempt to access the old flags? Like "the xxxxx_enabled flag is deprecated and will be removed in a future release. Please use xxxxx.available instead."

Comment (leofang) [reply cupy/cuda/__init__.py:29]: This question for `CUB_DISABLED` is related: https://github.com/cupy/cupy/pull/3562#issuecomment-657793261

Comment (kmaehashi) [reply cupy/cuda/__init__.py:29]: As for `CUB_DISABLED`, I think we don't have to emit depreaction as it was undocumented.

Comment (kmaehashi) [reply cupy/cuda/__init__.py]: Added DeprecationWarnings.

File: cupy/cuda/thrust.pyx (hunk 1 above)
Comment (kmaehashi) [reply cupy/cuda/thrust.pyx]: Fixed!

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
@kmaehashi Could you fix also `tests/cupy_tests/cuda_tests/test_cutensor.py` to use `cutensor.available`?

**Comment by asi1024** [comment]
Please resolve conflicts and fix also `cupyx/runtime.py`.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Could you check failure?

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
Could you resolve conflicts?

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
`TestCuTensorReduceNumpyArrayCreation` in `test_cutensor.py` was removed in #3744, but maybe restored in this PR?

**Comment by kmaehashi** [comment]
Sorry, I made a mistake when resolving conflict.
pfnCI, test this please.

**Comment by asi1024** [comment]
LGTM.

**Review by asi1024 — APPROVED** [review]

**Pull request merged**